### ISSUE TITLE
relative path issue works with windows native tool (cross compiler for example)

### ISSUE THREAD

**Issue by fchai** [msys2/MSYS2-packages#2126]
assume there are 3 windows drive:

C:\
D:\

there is a dir named "del" in "D:\".
there is a file named "d.txt" in "D:\".

In msys2, when I working in "/d/del/", the path of "../../d/d.txt" should works.

But this path works only for msys2 tools, for example "ls", "cat", "vim" etc. Command for example: "cat ../../d/d.txt".

Not works for windows tools, for example "notepad.exe", "cross tool compilers". Command for example: "notepad ../../d/d.txt", tools will complain the file not found.

**Comment by lazka** [comment]
You can't use a relative path that goes on level above the root drive, since there is nothing like this on Windows.

```
$ echo $(cygpath -m /d/del/)../../d/d.txt
D:/del/../../d/d.txt
```

**Comment by fchai** [comment]
> You can't use a relative path that goes on level above the root drive, since there is nothing like this on Windows.
> 
> ```
> $ echo $(cygpath -m /d/del/)../../d/d.txt
> D:/del/../../d/d.txt
> ```

Even there is no upper lever for root drive in Windows, but msys2 just simulated a root "/", and Windows drive like "X:/" is just mounted at "/x". So relative path like "../../../../d/aaa/bbb/source.c" should works. In fact, such path works with msys2 tools, for example, this is the result of "cat ../../../../d/aaa/bbb/source.c":

```
fchai@/d/aaa/ccc$ cat ../../../../d/aaa/bbb/source.c
#include <stdio.h>

int main(int argc, char * argv[])
{
    printf("Hello world\n");

    return 0;
}
```

By replacing "cat" with a Windows native tool or command in the command (for example, notepad.exe) will lead to error.

**Comment by lazka** [comment]
It simulates this "/" in the msys2 tools, the Windows tools/kernel know nothing about it.

**Comment by fchai** [comment]
msys2 just mounted Windows drive like "X:\" at "/x/", so the path "../../../d/aaa/bbb/source.c" should work.

But it works only for msys2 tools, not for Windows native tools because all relative path passed to Windows native tools is not converted.

**Comment by fchai** [comment]
> It simulates this "/" in the msys2 tools, the Windows tools/kernel know nothing about it.

You are right.

So msys2 should convert "../../../d/aaa/bbb/source.c" to the correct absolute path before invoking Windows tools. Just like "notepad /d/aaa/bbb/source.c"

Commands like "notepad /d/aaa/bbb/source.c" works fine because *nix absulote path is converted to Windows path. But msys2 does nothing against commands like "notepad ../../../d/aaa/bbb/source.c".

**Comment by lazka** [comment]
From what I understand the path conversion doesn't access/check anything on disk, it just converts based on the path content, so this is somewhat expected.

**Comment by fchai** [comment]
> From what I understand the path conversion doesn't access/check anything on disk, it just converts based on the path content, so this is somewhat expected.

IMO, `notepad ../../../d/aaa/bbb/source.c` should works just like `notepad /d/aaa/bbb/source.c`, because msys2 has nice path convertion.

**Comment by fchai** [comment]
I'm not sure if the msys2 path convertion is done by `https://github.com/msys2/path_convert/blob/master/src/path_conv.cpp`, if the answer is "Yes", from the line of 503 ("https://github.com/msys2/path_convert/blob/master/src/path_conv.cpp#L503") we can see: relative path is not processed.

**Comment by fchai** [comment]
> From what I understand the path conversion doesn't access/check anything on disk, it just converts based on the path content, so this is somewhat expected.

Yes, the convertion is just based on the literal path string, has no interactive with the mount pointers...

Not sure is it possible fix this issue?

**Comment by fchai** [comment]
> > From what I understand the path conversion doesn't access/check anything on disk, it just converts based on the path content, so this is somewhat expected.
> 
> Yes, the convertion is just based on the literal path string, has no interactive with the mount pointers...
> 
> Not sure is it possible to fix this issue?

IMO, it is possible to fix this issue just from the literal path string:

1) "/x/" => "X:/"
2) "../../../" => "/"
3) "../../../d/aaa/bbb/source.c" => "../../../" + "d/aaa/bbb/source.c"
4) "../../../d/aaa/bbb/source.c" => "/d/aaa/bbb/source.c"
5) "/d/aaa/bbb/source.c" => "D:/aaa/bbb/source.c"

**Comment by mingwandroid** [comment]
MSYS2 should never convert relative paths. It would cause absolute mayhem!

**Comment by fchai** [comment]
> MSYS2 should never convert relative paths. It would cause absolute mayhem!

Is it possible only convert relative path that first touch root like those mentioned in this issue ("../../../d/aaa/bbb/source.c")?
